### PR TITLE
Enhance/reportbackup

### PR DESCRIPTION
**Pull request by amitmalav**
* Added new app (report) for reporting purposes
* Added new request route `/report` dedicated for reporting
* Added logic for having configurable consecutive no backup days limit

Now `apps.js` file is only used in unit tests

### CODE REVIEW

File: ExpressApp.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+const express = require('express');
+const morgan = require('morgan');
+const bodyParser = require('body-parser');
+const lib = require('./lib');
+const config = lib.config;
+const logger = lib.logger;
+const middleware = lib.middleware;
+const connectTimeout = require('connect-timeout');
+
+class ExpressApp {
```
Comment (muralisc) [inline ExpressApp.js]: Use name corresponding to our use case ? like `ReportApp` ??
And rename across the file.

File: ReportServer.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+console.log('Starting Service Fabrik...');
```
Comment (muralisc) [inline ReportServer.js]: Starting Service Fabrik report app ??

File: lib/reports/BackupReportManager.js [hunk 3]
```diff
@@ -57,18 +55,24 @@ class BackupReportManager {
...
+        let windowStartIndex = 0; // start index of consecutive no backup days window
+        let windowEndIndex = 0; // end index of consecutive no backup days window
+        while (windowEndIndex < noBackupDays.length) {
+          while (
+            windowEndIndex + 1 < noBackupDays.length &&
+            noBackupDays[windowEndIndex].valueOf() === moment(noBackupDays[windowEndIndex + 1]).subtract(1, 'days').toDate().valueOf()
+          ) {
+            windowEndIndex++;
+          }
+          threeConsecutiveBackupFailureCount += (windowEndIndex - windowStartIndex + 1) >= maxAllowedNoBackupDaysInRow ? (windowEndIndex - windowStartIndex + 2 - maxAllowedNoBackupDaysInRow) : 0;
```
Comment (kamath-prasad) [inline lib/reports/BackupReportManager.js]: can you rename these variables now 

- `threeConsecutiveBackupFailureCount` to `consecutiveBackupFailureCount`
- `failedBackups3daysInARow` to `failedBackupsNdaysInARow` ?

**Comment by amitmalav** [comment]
Link to bosh release PR: https://github.com/cloudfoundry-incubator/service-fabrik-boshrelease/pull/38

File: ServerUtil.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    const port = app.get('port');
+    const title = app.get('title');
+    const ssl = app.get('ssl');
+    const type = app.get('type');
+    const server = ssl ? https.createServer(ssl, app) : http.createServer(app);
+    server.on('error', onerror);
+    server.on('listening', onlistening);
+    server.listen(port);
+
+    function onerror(err) {
```
Comment (kamath-prasad) [inline ServerUtil.js]: can you modify all the loggers in this function to concatenate in the standard way we do in other places ?

Like - https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/136/files#diff-839d93c5f783939b13c6ae0dee13f50bR41

File: lib/reports/BackupReportManager.js [hunk 5]
```diff
@@ -57,25 +55,31 @@ class BackupReportManager {
...
+          windowStartIndex = windowEndIndex;
+        }
         const summary = {
           noBackupDays: noBackupDays,
           backupsTriggerred: triggeredBackups.length,
           backupsSucceeded: backupSuccessCount,
           backupFailed: (resultBackups.length - backupSuccessCount),
           failedBackupGuids: resultBackups.slice(backupSuccessCount),
-          failedBackups3daysInARow: threeConsecutiveBackupFailureCount
+          failedBackupsNdaysInARow: consecutiveBackupFailureCount
```
[the remark this answers is not in the document]
Comment (amitmalav) [reply lib/reports/BackupReportManager.js]: This is count of instances when there was no backup taken for 3(N) consecutive days {3 consecutive no backup days}

File: lib/routes/index.js [hunk 6]
```diff
@@ -3,4 +3,5 @@
 exports.cf = require('./cf');
 exports.manage = require('./manage');
 exports.admin = require('./admin');
-exports.api = require('./api');
\ No newline at end of file
+exports.api = require('./api');
+exports.report = require('./report');
```
[the remark this answers is not in the document]
Comment (amitmalav) [reply lib/routes/index.js]: Changed it to `/admin/report`

File: ServerUtil.js (hunk 4 above)
Comment (amitmalav) [reply ServerUtil.js]: Done.

File: lib/reports/BackupReportManager.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (amitmalav) [reply lib/reports/BackupReportManager.js]: yeah, agree. will change.

File: lib/reports/BackupReportManager.js [hunk 7]
```diff
@@ -57,25 +55,31 @@ class BackupReportManager {
...
+          ) {
+            windowEndIndex++;
+          }
+          consecutiveBackupFailureCount += (windowEndIndex - windowStartIndex + 1) >= maxAllowedNoBackupDaysInRow ? (windowEndIndex - windowStartIndex + 2 - maxAllowedNoBackupDaysInRow) : 0;
+          windowEndIndex++;
+          windowStartIndex = windowEndIndex;
+        }
         const summary = {
           noBackupDays: noBackupDays,
           backupsTriggerred: triggeredBackups.length,
```
[the remark this answers is not in the document]
Comment (amitmalav) [reply lib/reports/BackupReportManager.js]: I felt that `backupsTriggerred` was intuitive enough to suggest it's a number. Shall change to `NumberOfBackupsTriggerred`  if it isn't so.

[the remark this answers is not in the document]
Comment (subhankarc) [reply lib/reports/BackupReportManager.js]: Ok keep it as is then.

File: FabrikApp.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+const morgan = require('morgan');
+const bodyParser = require('body-parser');
+const lib = require('./lib');
+const config = lib.config;
+const logger = lib.logger;
+const middleware = lib.middleware;
+const connectTimeout = require('connect-timeout');
+
+class FabrikApp {
+  static createApp(type, addRoutes) {
```
Comment (jagadish-kb) [inline FabrikApp.js]: rename this to `create`

File: HttpServer.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+const http = require('http');
+const https = require('https');
+const pubsub = require('pubsub-js');
+const lib = require('./lib');
+const CONST = require('./lib/constants');
+const logger = lib.logger;
+
+class HttpServer {
+  static startServer(app) {
```
Comment (jagadish-kb) [inline HttpServer.js]: rename this to `start`

File: ReportServer.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  app.get('/', (req, res) => {
+    res.render('index', {
+      title: app.get('title')
+    });
+  });
+  app.use('/admin/report', routes.report);
+});
+
+HttpServer.startServer(report);
+process.on('SIGTERM', HttpServer.notifyShutDown);
```
Comment (jagadish-kb) [inline ReportServer.js]: can these event handlers move over into `HttpServer.handleShutdown` & invoked from within `start` method. This is repetitive code which can be part of `HttpServer`

File: apps.js [hunk 11]
```diff
@@ -41,63 +32,22 @@ const external = createApp('external', app => {
   app.use('/manage', routes.manage);
 });
 
+const report = FabrikApp.createApp('report', app => {
```
Comment (jagadish-kb) [inline apps.js]: is this required? where is `app.js` used? if this is not used wouldnt it be better to get rid of this file?

Comment (amitmalav) [reply apps.js]: It is used in unit tests, what we can do is to move it to test directory.

**Pull request merged**